src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs

```diff
@@ -392,5 +411,99 @@ public int GetArrayRank()
 #else
         ImmutableArray<TypeName> GetGenericArguments() => _genericArguments;
 #endif
+
+#if SYSTEM_REFLECTION_METADATA
+        /// <summary>
+        /// Creates a new <see cref="TypeName" /> object that represents current simple name with provided assembly name.
+        /// </summary>
+        /// <param name="assemblyName">Assembly name.</param>
+        /// <returns>Created simple name.</returns>
+        /// <exception cref="InvalidOperationException">The current type name is not simple.</exception>
+        public TypeName MakeSimpleTypeName(AssemblyNameInfo? assemblyName)
+        {
+            if (!IsSimple)
+            {
+                TypeNameParserHelpers.ThrowInvalidOperation_NotSimpleName(FullName);
+            }
+
+            TypeName? declaringType = IsNested
+                ? DeclaringType.MakeSimpleTypeName(assemblyName)
+                : null;
+
+            return new TypeName(fullName: FullName,
```

Comment:
`FullName` includes both the declared type name and the nested type name. It means that this creates non-linear amount of work for types nested multiple times (e.g. try it on something like `A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A`). That does not sound right. Should this be just the name for nested types?

Reply:
Also, is the parser missing a node count check for nested types? I would expect `A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A+A` to fail to parse with default  `TypeNameParseOptions.MaxNodes`.

Reply:
> FullName includes both the declared type name and the nested type name. It means that this creates non-linear amount of work for types nested multiple times

Great catch! I was able to solve it by passing `_fullName` and `_nestedNameLength` to the ctor. It's going to help us avoid any allocations (and keep the promise this API makes ;) )

https://github.com/dotnet/runtime/blob/b080dc2490dfd238d3671e4721308aef39ef14c3/src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs#L148-L153

Reply:
> Also, is the parser missing a node count check for nested types? 

That is true, but it's a separate bug. Will send PR shortly.